app/gui2/src/stores/imports.ts

```diff
@@ -0,0 +1,589 @@
+import { SuggestionDb } from '@/stores/suggestionDatabase'
+import {
+  SuggestionKind,
+  entryQn,
+  makeCon,
+  makeMethod,
+  makeModule,
+  makeStaticMethod,
+  makeType,
+  type SuggestionEntry,
+} from '@/stores/suggestionDatabase/entry'
+import { Ast, AstExtended } from '@/util/ast'
+import { GeneralOprApp } from '@/util/ast/opr'
+import {
+  qnFromSegments,
+  qnSplit,
+  tryIdentifier,
+  tryQualifiedName,
+  type Identifier,
+  type QualifiedName,
+} from '@/util/qualifiedName'
+import { unwrap } from '@/util/result'
+
+// ========================
+// === Imports analysis ===
+// ========================
+
+function parseIdent(ast: AstExtended): Identifier | null {
+  if (ast.isTree(Ast.Tree.Type.Ident) || ast.isToken(Ast.Token.Type.Ident)) {
+    const ident = tryIdentifier(ast.repr())
+    return ident.ok ? ident.value : null
+  } else {
+    return null
+  }
+}
+
+function parseIdents(ast: AstExtended): Identifier[] | null {
+  if (ast.isTree(Ast.Tree.Type.Ident) || ast.isToken(Ast.Token.Type.Ident)) {
+    const ident = tryIdentifier(ast.repr())
+    return ident.ok ? [ident.value] : null
+  } else if (ast.isTree(Ast.Tree.Type.OprApp)) {
+    const opr = new GeneralOprApp(ast)
+    const operands = opr.operandsOfLeftAssocOprChain(',')
+    return [...operands].flatMap((operand) => {
+      if (operand && operand.type === 'ast') {
+        const ident = parseIdent(operand.ast)
+        return ident != null ? [ident] : []
+      } else {
+        return []
+      }
+    })
+  } else {
+    return null
+  }
+}
+
+function parseQualifiedName(ast: AstExtended): QualifiedName | null {
+  if (ast.isTree(Ast.Tree.Type.Ident) || ast.isToken(Ast.Token.Type.Ident)) {
+    const name = tryQualifiedName(ast.repr())
+    return name.ok ? name.value : null
+  } else if (ast.isTree(Ast.Tree.Type.OprApp)) {
+    const opr = new GeneralOprApp(ast)
+    const operands = opr.operandsOfLeftAssocOprChain('.')
+    const idents = [...operands].flatMap((operand) => {
+      if (operand && operand.type === 'ast') {
+        const ident = parseIdent(operand.ast)
+        return ident != null ? [ident] : []
+      } else {
+        return []
+      }
+    })
```

Comment:
This implementation will silence any non-valid identifiers (so `Foo.(2 + 2).Bar` will be created as `Foo.Bar`). I'd prefer to return null instead.

Reply:
Actually, we won’t go there in the first place because the parser will not produce a valid AST tree here. But I implemented it as suggested just to make sure.